opentimelineio/algorithms/filter.py

```diff
@@ -0,0 +1,285 @@
+    first traversal order.
+
+    If unary function returns an object, that object will be inserted into the
+    hierarchy, replacing whichever object was passed in to it.  If it returns
+    None, that object will be pruned, and all of its children will be omitted
+    from iteration.  If the function returns a tuple, that tuple will be
+    inserted into the hierarchy.
+
+    If you have a track: [A,B,C]
+
+    If your unary function is:
+        def fn(thing):
+            if thing.name == B:
+                return thing' # some transformation of B
+            else:
+                return thing
+
+    filtered_items(track, fn) => [A,B',C]
+
+    If your unary function is:
+        def fn(thing):
+            if thing.name == B:
+                return None
+            else:
+                return thing
+
+    filtered_items(track, fn) => [A,C]
+
+    If your unary function is:
+        def fn(thing):
+            if thing.name == B:
+                return tuple(B_1,B_2,B_3)
+            else:
+                return thing
+
+    filtered_items(track, fn) => [A,B_1,B_2,B_3,C]
+    """
+
+    # deep copy everything
+    mutable_object = copy.deepcopy(input_object)
+
+    prune_list = set()
+
+    header_list = [mutable_object]
+
+    if isinstance(mutable_object, otio.schema.Timeline):
+        header_list.append(mutable_object.tracks)
+
+    iter_list = header_list + list(mutable_object.each_child())
+
+    for child in iter_list:
+        if (
+            _safe_parent(child) is not None
+            and _is_in(child.parent(), prune_list)
+        ):
+            prune_list.add(child)
+            continue
+
+        parent = None
+        child_index = None
+        if _safe_parent(child) is not None:
+            child_index = child.parent().index(child)
+            parent = child.parent()
+            del child.parent()[child_index]
+
+        # first try to prune
+        if (types_to_prune and _isinstance_in(child, types_to_prune)):
+            result = None
+        # then try to pass through
+        elif types_to_filter and not _isinstance_in(child, types_to_filter):
+            result = child
+        # finally call the user function
+        else:
+            result = unary_filter_fn(child)
+
+        if child is mutable_object:
+            mutable_object = result
+
+        if result is None:
+            prune_list.add(child)
+            continue
+
+        if type(result) is not tuple:
+            result = [result]
+
+        if parent is not None:
+            parent[child_index:child_index] = result
+
+    return mutable_object
+
+
+def _safe_parent(child):
+    if hasattr(child, 'parent'):
+        return child.parent()
+    return None
+
+
+def filtered_with_sequence_context(
+    input_object,
+    reduce_fn,
+    types_to_prune=None,
+    types_to_filter=None,
+):
+    """Filter input_object with reduce_fn to make a new copy of object.
```

Comment:
This one-line description is the same as the one for `filtered_items`. Maybe it could be differentiated by being something like: `Filter input_object with reduce_fn to make a new copy of object given the context of surrounding items.`